pkg/gomod/selector.go

```diff
@@ -15,90 +14,49 @@ var (
 )
 
 // Matcher matches a module name.
-type Matcher interface {
-	// Match returns true if the module name matches.
-	Match(modname string) bool
-}
-
-// MatcherFunc is a function that implements the Matcher interface.
-func MatcherFunc(fn matcherFunc) Matcher {
-	return fn
-}
-
-type matcherFunc func(modname string) bool
-
-func (fn matcherFunc) Match(modname string) bool {
-	return fn(modname)
-}
+type Matcher func(modname string) bool
 
-// Selector is a set of matchers that can be used to select a subset of modules.
-type Selector struct {
-	Includes []Matcher
-	Excludes []Matcher
-}
-
-// Select returns true if the module should be selected.
-func (s Selector) Select(modname string) bool {
-	selected := false
-	for _, m := range s.Includes {
-		if m.Match(modname) {
-			selected = true
-			break
+// CurrentModulesMatcher matches the current project modules.
+func CurrentModulesMatcher() (Matcher, error) {
+	os := gowork.RealSystem{}
+	knownMods := sets.NewString()
+	mods, err := gowork.List(os, os)
+	if err != nil {
+		if !errors.Is(err, gowork.ErrInvalidGowork) {
+			return nil, err
 		}
-	}
-	if !selected {
-		return false
-	}
-	for _, m := range s.Excludes {
-		if m.Match(modname) {
-			return false
+		m, err := gowork.Current(os, os)
+		if err != nil {
+			return nil, err
 		}
-	}
-	return true
-}
-
-// CurrentModulesMatcher matches the current project modules.
-type CurrentModulesMatcher struct {
-	once sync.Once
-	mods []gowork.Module
-	err  error
-}
-
-func (c *CurrentModulesMatcher) Match(modname string) bool {
-	c.once.Do(func() {
-		os := gowork.RealSystem{}
-		c.mods, c.err = gowork.List(os, os)
-		if c.err != nil && errors.Is(c.err, gowork.ErrInvalidGowork) {
-			var m *gowork.Module
-			m, c.err = gowork.Current(os, os)
-			if m != nil {
-				c.mods = []gowork.Module{*m}
-			}
+		if m != nil {
+			mods = []gowork.Module{*m}
 		}
-	})
-	if c.err != nil {
-		log.Fatal(c.err)
 	}
-	for _, m := range c.mods {
-		if m.Name == modname {
-			return true
-		}
+
+	for _, m := range mods {
+		knownMods.Insert(m.Name)
 	}
-	return false
+
+	return knownMods.Has, nil
 }
 
 // DefaultSelector returns a selector that includes modules with given domain,
 // but excludes the current project modules.
-func DefaultSelector(domain string) (Selector, error) {
+func DefaultSelector(domain string) (Matcher, error) {
 	domain = strings.TrimSpace(domain)
 	if len(domain) == 0 {
-		return Selector{}, ErrNoDomain
+		return func(string) bool { return true }, ErrNoDomain
+	}
+
+	currentModules, err := CurrentModulesMatcher()
```

Comment:
Calling `DefaultSelector(domain string) (Matcher, error)` will also do "the work", of looking for project modules, regardless if the given module matches the given domain name.

Reply:
Is it really important to defer the error handling into late in the game? It looked to me that we were only using `DefaultSelector` in the `RunE` commands for `check`, `float`, and `needs`, each of which invokes the `DefaultSelector` once.

Reply:
Same as https://github.com/knative/test-infra/pull/3704/files#r1107629477